src/java.base/share/classes/jdk/internal/lang/CaseFolding.java.template

```diff
@@ -0,0 +1,300 @@
+     * {@code toLowerCase} to both the pattern and the input string. This effectively 'close' the class for
+     * matching.
+     * <p>
+     * The family/char-properties are not "closed" and should remain unchanged. This is acceptable per RL1.5,
+     * if their behavior is clearly specified.
+     * <p>
+     * This method addresses that requirement for the "range" construct within in character class by computing
+     * the additional characters that should be included to close the range under simple case folding:
+     * <p>
+     * For each character in the input range {@code [start, end]} (inclusive), if the character has a simple
+     * case folding mapping in Unicode's CaseFolding.txt, the mapping is not a round-trip map, and the mapped
+     * character is not already in the range, then that mapped character (typically lowercase) is added to
+     * the expansion set.
+     * <p>
+     * This allows regex character class "range" implementation to use the returned expansion set to support
+     * additional case-insensitive matching, without duplicating characters already covered by the existing
+     * regex range implementation. The expectation is the matching is done using both the uppercase and
+     * lowercase forms of the input character, for example
+     *
+     * <pre>{@code
+     *
+     *     ch -> inRange(lower, Character.toUpperCase(ch), upper) ||
+     *           inRange(lower, Character.toLower(ch), upper) ||
+     *           additionalClosingCharacters.contains(Character.toUpperCase(ch)) ||
+     *           additionalClosingCharacters.contains(Character.toUpperCase(ch))
+     * }</pre>
+     *
+     * <p>
+     * @spec https://www.unicode.org/reports/tr18/#Simple_Loose_Matches
+     * @param start the starting code point of the character range
+     * @param end the ending code point of the character range
+     * @return a {@code int[]} containing the all simple case equivalents of characters in the range, excluding
+     *         those already in the range
+     */
+    public static int[] getClassRangeClosingCharacters(int start, int end) {
+        int[] expanded = new int[expanded_case_cps.length];
+        int off = 0;
+        for (int cp : expanded_case_cps) {
+            if (cp >= start && cp <= end) {
+                int folding = expanded_case_map.get(cp);
+                if (folding < start || folding > end) {
+                    expanded[off++] = folding;
+                }
+            }
+        }
+        return Arrays.copyOf(expanded, off);
+    }
+
+    private static final Map<Integer, Integer> expanded_case_map = Map.ofEntries(
+%%%Expanded_Case_Map_Entries
+    );
+
+    private static final int[] expanded_case_cps = expanded_case_map.keySet()
+      .stream()
+      .mapToInt(Integer::intValue)
+      .toArray();
+
+
+     private static class CaseFoldingEntry {
+        final int cp;
+        final char[] folding;
```

Comment:
Consider storing the folding as a int or long directly to avoid the overhead of small char arrays.
Arrange to be able to compare the whole replacement with another codePoint, etc.